### PR TITLE
Erick555/3.30

### PR DESCRIPTION
**Pull request by Erick555**
This is attempt to fix https://github.com/flathub/org.gimp.GIMP/pull/37

### CODE REVIEW

File: org.gimp.GIMP.WithBase.json [hunk 1]
```diff
@@ -4,7 +4,7 @@
     "base": "org.gimp.BaseApp",
     "base-version": "stable",
     "runtime": "org.gnome.Platform",
-    "runtime-version": "3.28",
+    "runtime-version": "3.30",
```
Comment (LeandroStanger) [inline org.gimp.GIMP.WithBase.json]: Update runtime to 3.32

Comment (Erick555) [reply org.gimp.GIMP.WithBase.json]: One step at time 😄

**Comment by Erick555** [comment]
Building webkit now works and it doesn't spam logs that much. Next we have to add missing deps.

**Comment by Jehan** [comment]
> Building webkit now works

Well the last automatic build failed. :-/

**Comment by Erick555** [comment]
@Jehan I undid the previous changes during commit squash. I'll fix it.

**Comment by Erick555** [comment]
@Jehan build successful, please test.

**Comment by Jehan** [comment]
@Erick555 Good job. I'll have a look this week. I have 2 questions from a very quick look:

1) Did the 3.30 platform remove json-c? It used to be available in the base build.

2) Why did you get rid of the generic build flags?

**Comment by Erick555** [comment]
> Did the 3.30 platform remove json-c? It used to be available in the base build.

It seems so. [Apps have to bundle it themselves](https://github.com/search?q=org%3Aflathub+json-c&type=Code). There is `json-glib` but not `json-c`.

> Why did you get rid of the generic build flags?

Do you mean [those](https://github.com/flathub/org.gimp.GIMP/pull/52/files#diff-b02bf331a0a68d2e7a4f18b14eb527bbL9)? Since move to freedesktop-sdk 18.08 as base for all runtimes, default set of buildflags is inherited from runtime. The [default set](https://gitlab.com/freedesktop-sdk/freedesktop-sdk/blob/freedesktop-sdk-extensions-18.08.28.2/project.conf#L71) is quite extensive and targeted at additional security. That means apps don't have to set some standard flags like `-O2 -g` themselves anymore unless there is need to compile a nightmare like old webkit 😄 .

After starting the app, there are some warnings like below so it's possible that some dependency is still missing. I'm not much familiar with gimp so you may try to figure it yourself. Feel free to push in this branch or borrow changes to your own.

```
GIMP-Warning: Bad interpreter referenced in interpreter file /app/lib/gimp/2.0/interpreters/pygimp.interp: /usr/bin/python2

GIMP-Warning: Bad interpreter referenced in interpreter file /app/lib/gimp/2.0/interpreters/pygimp.interp: /usr/bin/python2

GIMP-Warning: Bad interpreter referenced in interpreter file /app/lib/gimp/2.0/interpreters/pygimp.interp: /usr/bin/python2

GIMP-Warning: Bad interpreter referenced in interpreter file /app/lib/gimp/2.0/interpreters/pygimp.interp: python2

GIMP-Warning: Bad binary format string in interpreter file /app/lib/gimp/2.0/interpreters/pygimp.interp


(gimp-2.10:3): Gtk-WARNING **: 20:57:26.636: /app/lib/gtk-2.0/2.10.0/immodules/im-ibus.so: undefined symbol: ibus_input_context_set_client_commit_preedit

(gimp-2.10:3): Gtk-WARNING **: 20:57:26.636: Loading IM context type 'ibus' failed

(gimp-2.10:3): Gtk-WARNING **: 20:57:26.808: /app/lib/gtk-2.0/2.10.0/immodules/im-ibus.so: undefined symbol: ibus_input_context_set_client_commit_preedit

(gimp-2.10:3): Gtk-WARNING **: 20:57:26.808: Loading IM context type 'ibus' failed

(gimp-2.10:3): Gtk-WARNING **: 20:57:26.817: /app/lib/gtk-2.0/2.10.0/immodules/im-ibus.so: undefined symbol: ibus_input_context_set_client_commit_preedit

(gimp-2.10:3): Gtk-WARNING **: 20:57:26.817: Loading IM context type 'ibus' failed

(gimp-2.10:3): Gtk-WARNING **: 20:57:26.821: /app/lib/gtk-2.0/2.10.0/immodules/im-ibus.so: undefined symbol: ibus_input_context_set_client_commit_preedit
```

**Comment by Jehan** [comment]
Indeed I reproduce these warnings and they are very bad. We lost Python plug-in support (even though I checked the config.log and GIMP was properly configured with Python support), and ibus doesn't work anymore (I tested, I cannot write with a text input engine anymore, well I tested Japanese and Korean), which would make the flatpak unusable for non-Western people.

This is unfortunately non-releasable as-is. :-/

We will need to understand what is happening before updating.

**Comment by Erick555** [comment]
@Jehan so python2 isn't available in Platform anymore. It's only in Sdk so it's available at build time but not at runtime. Apps have to use [shared-modules](https://github.com/flathub/shared-modules/tree/master/python2.7) same as for gtk2. I still don't know about ibus (which is available in Platform).

**Comment by Jehan** [comment]
> so python2 isn't available in Platform anymore. It's only in Sdk so it's available at build time but not at runtime.

Oh that explains why I was so confused. I had a look in the flatpak environment (with a `flatpak run --user --devel ` command) and python was there, executable and I even tested it. That was because it was only in the SDK! Now I get it.

>  I still don't know about ibus (which is available in Platform).

As for the ibus issue, that's weird as well. The shared lib does define the symbol which is written as not defined. Run from within the flatpak:

```
$ readelf -Ws  /app/lib/gtk-2.0/2.10.0/immodules/im-ibus.so |grep ibus_input_context_set_client_commit_preedit
    12: 0000000000000000     0 FUNC    GLOBAL DEFAULT  UND ibus_input_context_set_client_commit_preedit
```

It is there and defined!

If anyone wants to diagnose this further, I added the Sdk.Debug and GIMP.Debug flatpak extensions, then ran gdb inside the flatpak env, and broke the execution on the first "undefined symbol" warning. Here is the backtrace:

```
(gimp-2.10:41): Gtk-WARNING **: 17:39:06.215: /app/lib/gtk-2.0/2.10.0/immodules/im-ibus.so: undefined symbol: ibus_input_context_set_client_commit_preedit

Thread 1 "gimp-2.10" received signal SIGTRAP, Trace/breakpoint trap.
0x00007ffff709c6a5 in _g_log_abort (breakpoint=1) at ../glib/gmessages.c:554
554	    G_BREAKPOINT ();
(gdb) bt
#0  0x00007ffff709c6a5 in _g_log_abort (breakpoint=1) at ../glib/gmessages.c:554
#1  0x00007ffff709d9dd in g_logv (log_domain=0x7ffff7db805b "Gtk", log_level=G_LOG_LEVEL_WARNING, format=<optimized out>, args=args@entry=0x7fffffffc870) at ../glib/gmessages.c:1371
#2  0x00007ffff709dbb3 in g_log (log_domain=log_domain@entry=0x7ffff7db805b "Gtk", log_level=log_level@entry=G_LOG_LEVEL_WARNING, format=format@entry=0x7ffff7e4cb92 "%s") at ../glib/gmessages.c:1413
#3  0x00007ffff7c0a632 in gtk_im_module_load (module=0x555558646980) at gtkimmodule.c:116
#4  0x00007ffff71a2621 in g_type_module_use (module=module@entry=0x555558646980) at ../gobject/gtypemodule.c:244
#5  0x00007ffff7c0ad02 in _gtk_im_module_create (context_id=<optimized out>) at gtkimmodule.c:612
#6  0x00007ffff7c0b706 in gtk_im_multicontext_get_slave (multicontext=0x5555585d7c90) at gtkimmulticontext.c:254
#7  0x00007ffff7c0ba76 in gtk_im_multicontext_get_preedit_string (context=<optimized out>, str=0x7fffffffca38, attrs=0x7fffffffca40, cursor_pos=0x0) at gtkimmulticontext.c:312
#8  0x00007ffff7c08863 in IA__gtk_im_context_get_preedit_string (context=0x5555585d7c90, str=str@entry=0x7fffffffca38, attrs=attrs@entry=0x7fffffffca40, cursor_pos=cursor_pos@entry=0x0) at gtkimcontext.c:447
#9  0x00007ffff7baffb5 in gtk_entry_create_layout (include_preedit=1, entry=0x5555585861a0) at gtkentry.c:5409
#10 gtk_entry_ensure_layout (entry=entry@entry=0x5555585861a0, include_preedit=include_preedit@entry=1) at gtkentry.c:5488
#11 0x00007ffff7bb0f54 in gtk_entry_get_cursor_locations (entry=entry@entry=0x5555585861a0, type=type@entry=CURSOR_STANDARD, strong_x=strong_x@entry=0x7fffffffcb14, weak_x=weak_x@entry=0x0) at gtkentry.c:5937
#12 0x00007ffff7bb17d1 in update_im_cursor_location (entry=0x5555585861a0) at gtkentry.c:5368
#13 recompute_idle_func (data=0x5555585861a0) at gtkentry.c:5368
#14 0x00007ffff7a498a8 in gdk_threads_dispatch (data=0x5555585e4100) at gdk.c:534
#15 0x00007ffff7096578 in g_main_dispatch (context=0x555555dcce10) at ../glib/gmain.c:3182
#16 g_main_context_dispatch (context=context@entry=0x555555dcce10) at ../glib/gmain.c:3847
#17 0x00007ffff7096968 in g_main_context_iterate (context=0x555555dcce10, block=block@entry=1, dispatch=dispatch@entry=1, self=<optimized out>) at ../glib/gmain.c:3920
#18 0x00007ffff7096c92 in g_main_loop_run (loop=0x555555e05b00) at ../glib/gmain.c:4116
#19 0x0000555555637cc7 in app_run (full_prog_name=<optimized out>, filenames=<optimized out>, alternate_system_gimprc=<optimized out>, alternate_gimprc=<optimized out>, session_name=<optimized out>, 
    batch_interpreter=0x0, batch_commands=0x0, as_new=0, no_interface=0, no_data=0, no_fonts=0, no_splash=0, be_verbose=0, use_shm=1, use_cpu_accel=1, console_messages=0, use_debug_handler=0, 
    show_playground=0, show_debug_menu=0, stack_trace_mode=GIMP_STACK_TRACE_NEVER, pdb_compat_mode=GIMP_PDB_COMPAT_ON, 
    backtrace_file=0x555555cd7320 "/home/jehan/.config/GIMP/2.10/CrashLog/GIMP-crash-1554219538.txt") at app.c:440
#20 0x0000555555637565 in main (argc=<optimized out>, argv=<optimized out>) at main.c:524
(gdb) 
```

Anyway this is a bit tiring. Trying to understand these kind of issues within a sandboxed environment is absolutely tiresome.

@Erick555 I will let you diagnose this one as well, I think. I feel like you are the best suited right now. ;-)

**Comment by Erick555** [comment]
@Jehan both [python2](https://github.com/flathub/org.gimp.GIMP/pull/52/commits/9333e8df3211558b699d802fffa099388a8eb679) and [ibus](https://github.com/flathub/org.gimp.GIMP/pull/52/commits/54d10d22cd62ae4a2eb6f13e4b5dc9c950ea98da) issues are fixed now. I don't see warning anymore while running gimp. Please test yourself

**Comment by Jehan** [comment]
Awesome job @Erick555. Just did a very quick test. IME typing works again within GIMP and I don't see any warning when running it. I'll want to have a closer look to all the changes soon, but this looks fine from afar.

**Comment by Jehan** [comment]
@Erick555:
Ok I reviewed every commit again one by one, and tested the build. It looks all good and fine.
If you wish to do the bump to the 3.32 runtime also, feel free. 😛

Maybe also you may have fun with #26. Without this additional dependency, some features are missing in the GIMP flatpak, but I couldn't manage to get it to build for all 4 build platforms.

In any case, very good job. Thanks!

**Pull request merged**